source/vir/src/split_expression.rs

```diff
@@ -0,0 +1,546 @@
+            });
+            let arg_map = arg_map;
+            let e1_replaced = subsitute_argument(e1, &arg_map)?;
+            let bnd_replaced: Bnd = match &bnd.x {
+                BndX::Let(binders) => {
+                    let mut new_binders: Vec<Binder<Exp>> = vec![];
+                    for old_b in &**binders {
+                        let new_b = BinderX {
+                            name: old_b.name.clone(),
+                            a: subsitute_argument(&old_b.a, &arg_map)?,
+                        };
+                        new_binders.push(Arc::new(new_b));
+                    }
+                    Spanned::new(bnd.span.clone(), BndX::Let(Arc::new(new_binders)))
+                }
+                BndX::Quant(quant, bndrs, _trigs) => Spanned::new(
+                    bnd.span.clone(),
+                    BndX::Quant(quant.clone(), bndrs.clone(), Arc::new(vec![])),
+                ),
+                _ => {
+                    return Err((
+                        exp.span.clone(),
+                        format!("Unsupported binder during subsitution"),
+                    ));
+                }
+            };
+            let new_exp = ExpX::Bind(bnd_replaced, e1_replaced);
+            SpannedTyped::new(&exp.span, &exp.typ, new_exp)
+        }
+        ExpX::Const(_) => exp.clone(),
+        _ => {
+            return Err((exp.span.clone(), format!("Unsupported expression during subsitution")));
+        }
+    };
+    Ok(result)
+}
+
+fn is_bool_type(t: &Typ) -> bool {
+    match &**t {
+        TypX::Bool => true,
+        TypX::Boxed(b) => is_bool_type(b),
+        _ => false,
+    }
+}
+
+fn is_same_type(t1: &Typ, t2: &Typ) -> bool {
+    match (&**t1, &**t2) {
+        (TypX::Bool, TypX::Bool)
+        | (TypX::Int(_), TypX::Int(_))
+        | (TypX::Tuple(_), TypX::Tuple(_))
+        | (TypX::Lambda(_, _), TypX::Lambda(_, _))
+        | (TypX::Datatype(_, _), TypX::Datatype(_, _))
+        | (TypX::TypParam(_), TypX::TypParam(_))
+        | (TypX::TypeId, TypX::TypeId)
+        | (TypX::Air(_), TypX::Air(_)) => true,
+        (TypX::Boxed(b1), TypX::Boxed(b2)) => is_same_type(b1, b2),
+        _ => false,
+    }
+}
+
+pub(crate) fn tr_inline_expression(
+    body_exp: &Exp,
+    params: &Params,
+    exps: &Exps,
+) -> Result<Exp, (Span, String)> {
+    match *body_exp.typ {
+        TypX::Bool => (),
+        _ => {
+            return Err((
+                body_exp.span.clone(),
+                "Skip inlining for non-boolean expreesion".to_string(),
+            ));
+        }
+    };
+    let mut arg_map: HashMap<Arc<String>, Exp> = HashMap::new();
+    let mut count = 0;
+    for param in &**params {
+        let exp_to_insert = &exps[count];
+        if !is_same_type(&param.x.typ, &exp_to_insert.typ) {
+            return Err((
+                exp_to_insert.span.clone(),
+                "Error: arg type mismatch during expression inlining".to_string(),
+            ));
```

Comment:
Same question, in what context does this error happen? Is it a response to some user error, or does this indicate some internal VIR error?

Reply:
I added this check just to be cautious, and I am not sure if this can be triggered.  If this is triggered, it would be some internal VIR error.

Reply:
okay that's what I expected. Can you change it to a panic or at least make it more clear it's an internal error?

As written, it looks like a user-level error which might be confusing if a user thinks it means they did something wrong.

Reply:
You are right, I think it is more like a `panic`.